tracee-rules/input.go

```diff
@@ -0,0 +1,70 @@
+package main
+
+import (
+	"bufio"
+	"encoding/gob"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"time"
+
+	"github.com/aquasecurity/tracee/tracee-rules/types"
+)
+
+func setupStdinSource(inputSource string) (chan types.Event, error) {
+	res := make(chan types.Event)
+	scanner := bufio.NewScanner(os.Stdin)
+	go func() {
+		for scanner.Scan() {
+			event := scanner.Bytes()
+			switch inputSource {
+			case "tracee":
+				var e types.TraceeEvent
+				err := json.Unmarshal(event, &e)
+				if err != nil {
+					log.Printf("invalid json in %s: %v", string(event), err)
+				}
+				res <- types.Event(e)
+			}
+		}
+		close(res)
+	}()
+	return res, nil
+}
+
+func setupTraceeSource(traceeFilePath string) (chan types.Event, error) {
+	fi, err := os.Stat(traceeFilePath)
+	if err != nil || !fi.Mode().IsRegular() {
+		return nil, fmt.Errorf("invalid Tracee input file: %s", traceeFilePath)
+	}
+	f, err := os.Open(traceeFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("invalid file: %s", traceeFilePath)
+	}
+	dec := gob.NewDecoder(f)
+	res := make(chan types.Event)
+	go func() {
+		for {
+			var event types.TraceeEvent
+			err := dec.Decode(&event)
+			if err != nil {
+				if err == io.EOF {
+					// ignore EOF becasue we assume events can keep streaming into the file
+					time.Sleep(time.Millisecond * 500)
```

Comment:
The number of events that can occur in 500 msec may be huge.
Although it is ok that rule match will happen with some delay,
Buffer (pipe?) may not be enough for all of the incoming events

Reply:
good point.. but this "buffer" is a user consideration, if they give us a real file on disk, or a pipe or whatever, this is how they control the "buffer" size. what can tracee do besides documenting this consideration? maybe just lowering this sleep duration but any value will have the same issue

Reply:
We should document this then.
A possible solution (which complicates things) is to use adative sleep times - start from 10ms and go up to 500ms if there is no activity - I'm not in favor for that solution though.
There is another problem, which is that the consumer (sigengine) may read an incomplete event, and an errors will happen at the gob decoder.
We can document both of these issues for now.
A proper solution should be based on some kind of synchronization mechanism, and not involve sleeps.
For example, when 1000 events are ready to be processed, tracee can signal the processor that it can now consume 1000 events

Reply:
ok. do you think we should lower the value for now?

Reply:
Yes, 100/200 ms should be less "dangerous" to lose events